qdglue/metrics/grid_based_metrics.py

```diff
@@ -0,0 +1,265 @@
+        assert (
+            np.prod(fitness_bounds.shape) == 1
+        ), "Only scalar fitnesses are supported for now"
+
+        self.fitness_bounds = fitness_bounds
+        self.num_points_ccdf = num_points_ccdf
+
+    def __call__(
+        self,
+        fitnesses: Fitness,
+        features: Feature,
+    ) -> DiscreteArchiveMetrics:
+        return self.get_metrics(fitnesses, features)
+
+    def get_metrics(
+        self,
+        fitnesses: Fitness,
+        features: Feature,
+    ) -> DiscreteArchiveMetrics:
+        self._check_validity_features(features)
+        self._check_validity_fitnesses(fitnesses)
+
+        # Extracting the best fitnesses per cell
+        # The empty cells do not appear in the best_fitnesses array
+        best_fitnesses = self._extract_best_fitnesses_per_cell(fitnesses, features)
+        best_fitnesses_non_negative = best_fitnesses - self.fitness_bounds.low
+        best_fitnesses_normalized = best_fitnesses_non_negative / (
+            self.fitness_bounds.high - self.fitness_bounds.low
+        )
+
+        # Computing the metrics
+        total_cells = self.num_cells
+
+        coverage = np.size(best_fitnesses) / total_cells
+        max_fitness = np.max(best_fitnesses)
+        qd_score = np.sum(best_fitnesses_non_negative).item()
+        qd_score_bound_norm = np.sum(best_fitnesses_normalized).item()
+
+        # Computing the Complementary Cumulative Distribution of Fitness
+        ccdf = self._compute_ccdf(best_fitnesses)
+
+        return DiscreteArchiveMetrics(
+            coverage=coverage,
+            qd_score_original=qd_score,
+            qd_score_bound_norm=qd_score_bound_norm,
+            max_fitness=max_fitness,
+            ccdf=ccdf,
+            discrete_archive_metrics_calculator=self,
+        )
+
+    def _compute_ccdf(
+        self,
+        best_fitnesses: Fitness,
+    ) -> np.ndarray:
+        histogram_fitnesses = np.histogram(
+            best_fitnesses.ravel(),
+            bins=self.num_points_ccdf - 1,
+            range=(self.fitness_bounds.low.item(), self.fitness_bounds.high.item()),
+        )[0]
+
+        histogram_fitnesses = np.append(histogram_fitnesses, 0)
+
+        # Complementary Cumulative Distribution of Fitness (done by summing the histogram from the end)
+        ccdf = np.cumsum(histogram_fitnesses[::-1])[::-1] / self.num_cells
+
+        return ccdf
+
+    @property
+    @abstractmethod
+    def num_cells(self) -> int:
+        """Number of cells in the grid."""
+
+    @abstractmethod
+    def _check_validity_features(self, features: Feature) -> None:
+        """Check validity of features."""
+
+    def _check_validity_fitnesses(self, fitnesses):
+        assert np.all(np.min(fitnesses, axis=0) >= self.fitness_bounds.low) and np.all(
+            np.max(fitnesses, axis=0) <= self.fitness_bounds.high
+        ), "Fitnesses must be within the fitness bounds"
+
+    @abstractmethod
+    def _extract_best_fitnesses_per_cell(
```

Comment:
It seems you're using inheritance here primarily as a way to reuse code, which means that child classes are left to implement a seemingly random set of abstract methods (compare: just implementing forward() in pytorch).

The design depends a lot on our goal here. If we just want to make calculators that the user can use, then this works fine. However, if we want the user to be able to implement their own calculators, we'll have to think a lot more about how these classes should be designed. We may want to think of how we can use composition instead of inheritance here, for instance.